### PR TITLE
Add value_set commands

### PR DESCRIPTION
**Pull request by kmateuszssak**

### CODE REVIEW

**Comment by osinstom** [comment]
@kmateuszssak don't forget about updating the docs

File: CLI/common.h [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include "../include/psabpf.h"
 #include <arpa/inet.h>
 #include <ctype.h>
+#include <jansson.h>
```
Comment (tatry) [inline CLI/common.h:22]: This header is required in this file?

File: README.md [hunk 2]
```diff
@@ -174,3 +174,21 @@ psabpf-ctl table add pipe <ID> <TABLE> action <ACTION> key <KEY> data <DATA> [pr
...
+```shell
+psabpf-ctl value_set delete pipe ID VALUE_SET_NAME value DATA
+```
+List all elements in the value_set:
+```shell
+psabpf-ctl value_set get pipe ID VALUE_SET_NAME
+```
+- ID - ID of the pipeline, natural number.
+- VALUE_SET_NAME - name of the value_set.
+- DATA - the value_set value. Values can be specified in binary, octal, decimal or hexadecimal format.
```
Comment (tatry) [inline README.md:194]: In `README.md` only the most important commands are described, in my opinion parser value_set is not one of these commands. But this command should be added to the [command reference](https://github.com/P4-Research/psabpf/blob/master/docs/command%20reference.md).

File: main.c [hunk 3]
```diff
@@ -71,7 +72,8 @@ static int do_help(int argc, char **argv)
             "                   meter |\n"
             "                   digest |\n"
             "                   counter |\n"
-            "                   register }\n"
+            "                   register |\n"
+            "                   value_set }\n"
```
Comment (tatry) [inline main.c:76]: Maybe `value-set` to be conformant with other objects?

File: lib/psabpf_value_set.c [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        free(value->raw_data);
+    value->raw_data = NULL;
+}
+
+static int parse_key_type(psabpf_value_set_context_t *ctx)
+{
+    uint32_t type_id = 0;
+    if (ctx->btf_metadata.btf != NULL) {
+        type_id = psabtf_get_member_type_id_by_name(ctx->btf_metadata.btf, ctx->set_map.btf_type_id, "key");
+    }
```
Comment (tatry) [inline lib/psabpf_value_set.c:67]:
```suggestion
    uint32_t type_id = psabtf_get_member_type_id_by_name(ctx->btf_metadata.btf, ctx->set_map.btf_type_id, "key");
```
`psabtf_get_member_type_id_by_name` should correctly handle NULL BTF pointer now.

File: include/psabpf_value_set.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+void psabpf_value_set_context_init(psabpf_value_set_context_t *ctx);
+void psabpf_value_set_context_free(psabpf_value_set_context_t *ctx);
+int psabpf_value_set_context_name(psabpf_context_t *psabpf_ctx, psabpf_value_set_context_t *ctx, const char *name);
+
+void psabpf_value_set_init(psabpf_value_set_t *value);
+void psabpf_value_set_free(psabpf_value_set_t *value);
+psabpf_value_set_t * psabpf_value_set_get_next(psabpf_value_set_context_t *ctx);
+
+int psabpf_value_set_set_value(psabpf_value_set_t *value, const void *data, size_t data_len);
+psabpf_struct_field_t * psabpf_value_set_get_next_value_field(psabpf_value_set_context_t *ctx, psabpf_value_set_t *entry);
```
Comment (tatry) [inline include/psabpf_value_set.h:47]: For now it works for exact match. But what if we implement other match kinds (ternary and LPM) for pvs? For this reason I suggest to reuse `match_key` API (and implementation if possible) from tables.

File: lib/psabpf_value_set.c [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+    int ret = open_bpf_map(psabpf_ctx, name, &ctx->btf_metadata, &ctx->set_map);
+    if (ret != NO_ERROR) {
+        fprintf(stderr, "couldn't open a value_set %s\n", name);
+        return ret;
+    }
+
+    if (parse_key_type(ctx) != NO_ERROR) {
+        fprintf(stderr, "%s: couldn't get key BTF info of a value_set instance\n", name);
```
Comment (tatry) [inline lib/psabpf_value_set.c:87]:
```suggestion
        fprintf(stderr, "%s: couldn't parse structure of a value_set instance\n", name);
```

File: CLI/value_set.c [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    (void) argc; (void) argv;
+    fprintf(stderr,
+            "Usage: %1$s value_set get pipe ID VALUE_SET_NAME\n"
+            "       %1$s value_set insert pipe ID VALUE_SET_NAME value DATA\n"
+            "       %1$s value_set delete pipe ID VALUE_SET_NAME value DATA\n"
+            "",
+            program_name);
+
+    return NO_ERROR;
+}
```
Comment (tatry) [inline CLI/value_set.c:241]: Missing new line ant the end of file.

**Review by tatry — COMMENTED** [review]
Few questions from my side

File: include/psabpf_value_set.h (hunk 5 above)
Comment (kmateuszssak) [reply include/psabpf_value_set.h:47]: This is too much work to do. And a work that cannot be tested anyway. Let's stay with that for now.

File: docs/command reference.md [hunk 8]
```diff
@@ -133,3 +133,11 @@ psabpf-ctl register set pipe ID REGISTER_NAME index DATA value REGISTER_VALUE
 
 REGISTER_VALUE := { DATA }
 ```
+
+# Value set
+
+```shell
+psabpf-ctl value-set insert pipe ID VALUE_SET_NAME value DATA
```
Comment (osinstom) [inline docs/command reference.md:140]: Any reason to change convention from `add` to `insert` ? We use `add` for insertion in other commands. I'd keep it consistent.

File: include/psabpf_value_set.h (hunk 5 above)
Comment (osinstom) [reply include/psabpf_value_set.h:47]: How much work would it really be? If we can invest more time now to make it more extensible/cleaner, I'd go for it. We don't have any deadlines, etc. so we can spend more time on this, if needed.

**Review by tatry — CHANGES_REQUESTED** [review]
One important thing from my side

File: include/psabpf.h [hunk 9]
```diff
@@ -500,6 +500,7 @@ int psabpf_table_entry_add(psabpf_table_entry_ctx_t *ctx, psabpf_table_entry_t *
 int psabpf_table_entry_update(psabpf_table_entry_ctx_t *ctx, psabpf_table_entry_t *entry);
 int psabpf_table_entry_del(psabpf_table_entry_ctx_t *ctx, psabpf_table_entry_t *entry);
 int psabpf_table_entry_get(psabpf_table_entry_ctx_t *ctx, psabpf_table_entry_t *entry);
+int psabpf_table_entry_get_next_key(psabpf_table_entry_ctx_t *ctx);
```
[the remark this answers is not in the document]
Comment (kmateuszssak) [reply include/psabpf.h:503]: `psabpf_table_entry_get_next_key` does inside `bpf_map_get_next_key`. I moved the old begining part of `psabpf_table_entry_get_next` to this new function. Now value-set get command can read only keys not values.

**Pull request merged**